driver-core/src/main/java/com/datastax/driver/core/EventDebouncer.java

```diff
@@ -0,0 +1,208 @@
+    }
+
+    void start() {
+        logger.trace("Starting {} debouncer...", name);
+        started.countDown();
+        logger.trace("{} debouncer started", name);
+        eventsLock.lock();
+        try {
+            if (!events.isEmpty()) {
+                logger.trace("{} debouncer: events were accumulated before the debouncer started: scheduling delivery now", name);
+                scheduleNewDelivery();
+            }
+        } finally {
+            eventsLock.unlock();
+        }
+    }
+
+    void stop() {
+        logger.trace("Stopping {} debouncer...", name);
+        stopped = true;
+        while (true) {
+            DeliveryAttempt previous = cancelPendingDelivery();
+            if (pendingDelivery.compareAndSet(previous, null)) {
+                break;
+            }
+        }
+        logger.trace("{} debouncer stopped", name);
+    }
+
+    void eventReceived(T event) {
+        if(stopped) {
+            logger.trace("{} debouncer is stopped, rejecting event: {}", name, event);
+            return;
+        }
+        checkNotNull(event);
+        logger.trace("{} debouncer: event received {}", name, event);
+        enqueueEvent(event);
+        scheduleNewDelivery();
+    }
+
+    private void enqueueEvent(T event) {
+        eventsLock.lock();
+        try {
+            while (!events.offer(event)) {
+                logger.trace("{} debouncer: event queue is full, blocking the caller and attempting to deliver immediately", name);
+                try {
+                    // if the queue is full and the debouncer is not ready,
+                    // we have no other choice than to block and wait
+                    started.await();
+                    // this can race with a scheduled delivery,
+                    // but in the worse case the number of delivered events will be zero
+                    // which is harmless
+                    deliverEvents();
+                } catch (InterruptedException e) {
+                    logger.warn("Interrupted while attempting to deliver an event");
+                    Thread.currentThread().interrupt();
+                }
+            }
+        } finally {
+            eventsLock.unlock();
+        }
+    }
+
+    private DeliveryAttempt cancelPendingDelivery() {
+        DeliveryAttempt previous = pendingDelivery.get();
+        if (previous != null)
+            previous.cancel();
+        return previous;
+    }
+
+    private void scheduleNewDelivery() {
+        while (isRunning()) {
+            DeliveryAttempt previous = cancelPendingDelivery();
+            DeliveryAttempt next = new DeliveryAttempt();
+            if (pendingDelivery.compareAndSet(previous, next)) {
+                next.schedule();
+                break;
+            }
+        }
+    }
+
+    private boolean isRunning() {
+        return started.getCount() == 0 && !stopped;
+    }
+
+    ListenableFuture<?> deliverEvents() {
```

Comment:
The normal case is to have this method called by a thread in the scheduled executor, when the timer times out. 
But the following threads are also likely to execute this method in rare situations:

* IO thread (nio worker) in Cluster.handle(Response), if the queue is full (back pressure - note that the calling thread only executes the queue drain, once the events are drained they are processed in the main executor)
* Client thread or cluster-reconnection pool in ControlConnection.tryConnect(Host, boolean) (only happens at initial connection attempt, and in this method already blocks and everything is done synchronously)
* Cluster main executor in maybeRefreshSchemaAndSignal() - I wonder if we could have a deadlock here?